matrix_content_scanner/scanner/scanner.py

```diff
@@ -390,3 +404,51 @@ def _run_scan(self, file_name: str) -> int:
         except subprocess.CalledProcessError as e:
             logger.info("Scan failed with exit code %d: %s", e.returncode, e.stderr)
             return e.returncode
+
+    def _check_mimetype(
+        self,
+        media_content: bytes,
+        content_type_header: str,
+        encrypted: bool,
+    ) -> None:
+        """Reads the MIME type of the provided bytes, and checks that:
+        * it matches with the Content-Type header that was received when downloading this
+            file (if the media isn't encrypted, since otherwise the Content-Type header
+            is always 'application/octet-stream')
+        * files with this MIME type are allowed (if an allow list is provided in the
+            configuration)
+        Args:
+            media_content: The file's content. If the file is encrypted, this is its
+                decrypted content.
+            content_type_header: The value of the Content-Type header received when
+                downloading the file.
+            encrypted: Whether the file was encrypted (in which case we don't want to
+                check that its MIME type matches with the Content-Type header).
+        Raises:
+            FileDirtyError if one of the checks fail.
+        """
+        mimetype = magic.mimetype(media_content)
+        logger.info("MIME type for file is %s", mimetype)
+
+        # Check if the MIME type is matching the one that's expected, but only if the file
+        # is not encrypted (because otherwise we'll always have 'application/octet-stream'
+        # in the Content-Type header regardless of the actual MIME type of the file).
+        if encrypted is False and mimetype != content_type_header:
+            logger.error(
+                "Mismatching MIME type (%s) and Content-Type header (%s)",
+                mimetype,
+                content_type_header,
+            )
+            raise FileDirtyError("File type not supported")
```

Comment:
There is a chance that matrix clients may label uploads with a different MIME type to what `magic` detects. I think it's worth doing some manual testing, or keeping an eye out for bug reports once deployed.

Reply:
> or keeping an eye out for bug reports once deployed

Yes definitely. Worth noting that this would also be an issue with the "old" content scanner since it also supports this feature (not via libmagic but via https://www.npmjs.com/package/file-type, which afaict detects MIME types the same way).